### PR TITLE
feat(blade): RadioGroup - PR2

### PR DESCRIPTION
**Pull request by anuraghazra**
- Added RadioGroup/useRadioGroup

Prev: #605 
Next: #615 

-----

PR Stack: 

1. #605
2. #607 (Current)
3. #615
4. #629
5. #631

### CODE REVIEW

**Review by divyanshu013 — COMMENTED** [review]
Looks good overall, some comments.

File: packages/blade/src/components/Radio/RadioGroup/RadioGroup.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import { RadioGroupField } from './RadioGroupField';
+import { useRadioGroup } from './useRadioGroup';
+import Box from '~components/Box';
+import { FormHint, FormLabel } from '~components/Form';
+
+type RadioGroupProps = {
+  /**
+   * Accepts multiple radios as children
+   */
+  children: React.ReactNode;
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Radio/RadioGroup/RadioGroup.tsx:13]: Yeah unfortunate but specifying radio type will not change anything. JSX children's cant be restricted.

File: packages/blade/src/components/Radio/RadioGroup/RadioGroup.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+  const showError = validationState === 'error' && errorText;
+  const showHelpText = !showError && helpText;
+  const accessibillityText = `,${showError ? errorText : ''} ${showHelpText ? helpText : ''}`;
+
+  return (
+    <RadioGroupProvider value={contextValue}>
+      <RadioGroupField labelledBy={ids.labelId}>
+        <FormLabel
+          as="span"
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Radio/RadioGroup/RadioGroup.tsx:107]: It's Group nah... group doesn't need label the single radios have associated label. 

form Groups need fieldset & legend (which this span is equivalent to by providing ID)

**Review by divyanshu013 — APPROVED** [review]
Looks good to me.

File: packages/blade/src/components/Radio/RadioGroup/RadioGroup.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+};
+
+const RadioGroup = ({
+  children,
+  label,
+  helpText,
+  isDisabled,
+  neccessityIndicator = 'none',
+  labelPosition = 'top',
+  validationState,
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Radio/RadioGroup/RadioGroup.tsx:82]: Shouldn't we give a default value as `none` here?

File: packages/blade/src/components/Radio/RadioGroup/RadioGroup.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+   * (Useful for form submission).
+   */
+  name?: string;
+};
+
+const RadioGroup = ({
+  children,
+  label,
+  helpText,
+  isDisabled,
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Radio/RadioGroup/RadioGroup.tsx:79]: Lets give default value as `false`?

File: packages/blade/src/components/Radio/RadioGroup/RadioGroupField.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import { useRadioGroupContext } from './RadioContext';
+import { makeAccessible } from '~utils';
+import Box from '~components/Box';
+
+type RadioGroupFieldProps = {
+  children: React.ReactNode;
+  labelledBy: string;
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Radio/RadioGroup/RadioGroupField.tsx:8]: 🙈
```suggestion
  labeledBy: string;
```

File: packages/blade/src/components/Radio/RadioGroup/useRadioGroup.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
+/* eslint-disable @typescript-eslint/no-shadow */
+/* eslint-disable @typescript-eslint/prefer-nullish-coalescing */
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Radio/RadioGroup/useRadioGroup.ts:4]: Can't we just follow these rules?

File: packages/blade/src/components/Radio/RadioIcon.tsx [hunk 7]
```diff
@@ -4,7 +4,7 @@ import React from 'react';
 import { useTheme } from '~components/BladeProvider';
 import Svg from '~components/Icons/_Svg';
 import { getIn } from '~utils';
-import { CheckboxIconWrapper } from '~components/Checkbox/CheckboxIcon/CheckboxIconWrapper';
+import { CheckboxIconWrapper as RadioIconWrapper } from '~components/Checkbox/CheckboxIcon/CheckboxIconWrapper';
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Radio/RadioIcon.tsx:7]: Lets move the `CheckboxIconWrapper` to a common place and use it for both Radio & Checkbox?

**Review by chaitanyadeorukhkar — CHANGES_REQUESTED** [review]

File: packages/blade/src/components/Radio/Radio.tsx [hunk 8]
```diff
@@ -1,10 +1,12 @@
...
 import React from 'react';
 import { useRadio } from './useRadio';
 import { RadioIcon } from './RadioIcon';
+import { useRadioGroupContext } from './RadioGroup/RadioContext';
+import { SelectorLabel } from '~components/Form/Selector/SelectorLabel.web';
 import Box from '~components/Box';
-import { CheckboxInput } from '~components/Checkbox/CheckboxInput';
-import { SelectorLabel } from '~components/Form/Selector/SelectorLabel';
 import { SelectorTitle } from '~components/Form/Selector/SelectorTitle';
+import { CheckboxInput as RadioInput } from '~components/Checkbox/CheckboxInput';
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Radio/Radio.tsx:9]: Yep everything related to this has been done in #631 after me and kamlesh had discussion.

File: packages/blade/src/components/Radio/RadioGroup/RadioGroup.tsx (hunk 3 above)
Comment (anuraghazra) [reply packages/blade/src/components/Radio/RadioGroup/RadioGroup.tsx:82]: Yeah should be fine.

File: packages/blade/src/components/Radio/RadioGroup/RadioGroupField.tsx (hunk 5 above)
Comment (anuraghazra) [reply packages/blade/src/components/Radio/RadioGroup/RadioGroupField.tsx:8]: Oi... I'm British Oi... 
HTML spec seem to follow aria-`labelledby`

image

File: packages/blade/src/components/Radio/RadioIcon.tsx (hunk 7 above)
Comment (anuraghazra) [reply packages/blade/src/components/Radio/RadioIcon.tsx:7]: Yep done in #631

File: packages/blade/src/components/Radio/Radio.tsx [hunk 9]
```diff
@@ -2,11 +2,12 @@
 import React from 'react';
 import { useRadio } from './useRadio';
 import { RadioIcon } from './RadioIcon';
+import { useRadioGroupContext } from './RadioGroup/RadioContext';
+import { SelectorLabel } from '~components/Form/Selector/SelectorLabel.web';
```
[the remark this answers is not in the document]
Comment (divyanshu013) [reply packages/blade/src/components/Radio/Radio.tsx]: Your eyes are as sharp as a linter 😄

[the remark this answers is not in the document]
Comment (divyanshu013) [reply packages/blade/src/components/Radio/Radio.tsx]: Sharper than eslint in this case :)

File: packages/blade/src/components/Radio/Radio.tsx [hunk 10]
```diff
@@ -86,27 +87,45 @@ const Radio = ({
   helpText,
   validationState,
 }: RadioProps): React.ReactElement => {
+  const groupProps = useRadioGroupContext();
+
+  const _validationState = validationState ?? groupProps?.validationState;
+  const _hasError = _validationState === 'error';
+  const _isDisabled = isDisabled ?? groupProps?.isDisabled;
+  const _name = groupProps?.name ?? name;
+  const _isChecked = groupProps?.state?.isChecked(value!) ?? isChecked;
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Radio/Radio.tsx:97]: Right. I've changed it in another PRs.

**Pull request merged**